### PR TITLE
Add `DiagonalReduction` transformation for `TensorNetwork`

### LINKED ISSUE #14: Implement "Diagonal reduction" transformation

### PR DESCRIPTION
**Pull request by jofrevalles**
### Summary
This PR addresses the issue #14 (resolves #14) by introducing the `DiagonalReduction` transformation in the `transform!` function for `TensorNetwork`s. This transformation applies diagonal reduction (as defined [here](https://arxiv.org/pdf/2002.01935.pdf)) to simplify the tensor network while preserving its mathematical properties. The key idea is to minimize the number of indices by identifying tensors with a certain diagonal structure and collapsing these indices, potentially leading to computational efficiency.

In addition to the implementation of the transformation, this PR includes tests to ensure correctness and robustness of the new method.

### Example
```julia
julia> using Tenet

julia> data = [1.0 0.0; 0.0 1.0;;; 1.0 0.0; 0.0 1.0] # We can clearly see that the first and second indices are diagonal
2×2×2 Array{Float64, 3}:
[:, :, 1] =
 1.0  0.0
 0.0  1.0

[:, :, 2] =
 1.0  0.0
 0.0  1.0

julia> B = Tensor(rand(2, 2, 2), (:j, :l, :m))
2×2×2 Tensor{Float64, 3, Array{Float64, 3}}: ...

julia> tn = TensorNetwork([A, B])
TensorNetwork{Arbitrary}(#tensors=2, #inds=5)

julia> [labels(tensor) for tensor in tn.tensors]
2-element Vector{Tuple{Symbol, Symbol, Vararg{Symbol}}}:
 (:i, :k)
 (:i, :l, :m)
```
This example demonstrates how `DiagonalReduction` effectively reduces the dimension of the first tensor and updates the rest of the network accordingly.

### CODE REVIEW

File: src/Transformations.jl [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
 
 abstract type Transformation end
 
-transform(tn::TensorNetwork, transformations) = (tn = copy(tn); transform!(tn, transformations); return tn)
+transform(tn::TensorNetwork, transformations; kwargs...) = (tn = copy(tn); transform!(tn, transformations; kwargs...); return tn)
 
 function transform! end
 
-transform!(tn::TensorNetwork, transformation::Type{<:Transformation}) = transform!(tn, transformation())
+transform!(tn::TensorNetwork, transformation::Type{<:Transformation}; kwargs...) = transform!(tn, transformation(); kwargs...)
```
Comment (mofeing) [inline src/Transformations.jl:10]:
```suggestion
transform!(tn::TensorNetwork, transformation::Type{<:Transformation}; kwargs...) = transform!(tn, transformation(kwargs...))
```

File: src/Transformations.jl [hunk 2]
```diff
@@ -44,4 +45,62 @@ function transform!(tn::TensorNetwork, ::HyperindConverter)
     end
 end
 
-# TODO column reduction, diagonal reduction, rank simplification, split simplification
+struct DiagonalReduction <: Transformation end
+
+function transform!(tn::TensorNetwork, ::DiagonalReduction; output_inds=nothing, atol=1e-12)
+    output_inds = output_inds === nothing ? openinds(tn) : output_inds
```
Comment (mofeing) [inline src/Transformations.jl:51]:
```suggestion
function transform!(tn::TensorNetwork, config::DiagonalReduction)
    skip_inds = isempty(config.skip) ? openinds(tn) : config.skip
```

File: src/Transformations.jl [hunk 3]
```diff
@@ -44,4 +45,62 @@ function transform!(tn::TensorNetwork, ::HyperindConverter)
...
+        tensor = tn.tensors[idx]
+
+        diag_axes = find_diag_axes(parent(tensor), atol)
+
+        while !isempty(diag_axes) # loop over all diagonal axes
+            (i, j) = pop!(diag_axes)
+            ix_i, ix_j = labels(tensor)[i], labels(tensor)[j]
+
+            # do not reduce output indices
+            new, old = (ix_j in output_inds) ? ((ix_i in output_inds) ? continue : (ix_j, ix_i)) : (ix_i, ix_j)
```
Comment (mofeing) [inline src/Transformations.jl:65]:
```suggestion
            new, old = (ix_j in skip_inds) ? ((ix_i in skip_inds) ? continue : (ix_j, ix_i)) : (ix_i, ix_j)
```

File: src/Transformations.jl [hunk 4]
```diff
@@ -44,4 +45,62 @@ function transform!(tn::TensorNetwork, ::HyperindConverter)
...
+                    new_tensor = replace(other_tensor, old => new)
+                    tn.tensors[other_idx] = new_tensor
+                end
+            end
+
+            repeated_labels = replace(collect(labels(tensor)), old => new) |> join
+            removed_label = filter(l -> l != old, labels(tensor)) |> join
+
+            ein_code_str = "ein\"$repeated_labels->$removed_label\""
+            data = eval(Meta.parse("$(ein_code_str)($tensor)"))
```
Comment (mofeing) [inline src/Transformations.jl:80]: Don't call the compiler with `eval` and `Meta.parse`. This slows downs a lot. Use `EinCode` like I wrote in `contract`.

File: src/Transformations.jl [hunk 5]
```diff
@@ -1,16 +1,17 @@
 using DeltaArrays
+using OMEinsum
 
 abstract type Transformation end
 
-transform(tn::TensorNetwork, transformations) = (tn = copy(tn); transform!(tn, transformations); return tn)
+transform(tn::TensorNetwork, transformations; kwargs...) = (tn = copy(tn); transform!(tn, transformations; kwargs...); return tn)
```
[the remark this answers is not in the document]
Comment (jofrevalles) [reply src/Transformations.jl:6]: Got it!

File: src/Transformations.jl [hunk 6]
```diff
@@ -44,4 +45,62 @@ function transform!(tn::TensorNetwork, ::HyperindConverter)
     end
 end
 
-# TODO column reduction, diagonal reduction, rank simplification, split simplification
+struct DiagonalReduction <: Transformation end
```
[the remark this answers is not in the document]
Comment (jofrevalles) [reply src/Transformations.jl:48]: I didn't know that one existed, sorry

File: src/Transformations.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (mofeing) [reply src/Transformations.jl:6]:
```suggestion
transform(tn::TensorNetwork, transformations) = (tn = copy(tn); transform!(tn, transformations); return tn)
```

File: src/Transformations.jl [hunk 7]
```diff
@@ -44,4 +45,65 @@ function transform!(tn::TensorNetwork, ::HyperindConverter)
...
+
+function transform!(tn::TensorNetwork, config::DiagonalReduction)
+    skip_inds = isempty(config.skip) ? openinds(tn) : config.skip
+    queue = collect(keys(tn.tensors))
+
+    while !isempty(queue) # loop over all tensors
+        idx = pop!(queue)
+        tensor = tn.tensors[idx]
+
+        diag_axes = find_diag_axes(parent(tensor), atol)
```
Comment (mofeing) [inline src/Transformations.jl:61]:
```suggestion
        diag_axes = find_diag_axes(parent(tensor), config.atol)
```

File: src/Transformations.jl [hunk 8]
```diff
@@ -44,4 +45,65 @@ function transform!(tn::TensorNetwork, ::HyperindConverter)
...
+            repeated_labels = replace(collect(labels(tensor)), old => new) |> join
+            removed_label = filter(l -> l != old, labels(tensor)) |> join
+
+            ein_code_str = "ein\"$repeated_labels->$removed_label\""
+            data = eval(Meta.parse("$(ein_code_str)($tensor)"))
+            tn.tensors[idx] = Tensor(data, filter(l -> l != old, labels(tensor)))
+            delete!(tn.inds, old)
+
+            tensor = tn.tensors[idx]
+            diag_axes = find_diag_axes(parent(tensor), atol)
```
Comment (mofeing) [inline src/Transformations.jl:88]:
```suggestion
            diag_axes = find_diag_axes(parent(tensor), config.atol)
```

File: src/Transformations.jl [hunk 9]
```diff
@@ -44,4 +45,64 @@ function transform!(tn::TensorNetwork, ::HyperindConverter)
...
+                if old in labels(other_tensor)
+                    new_tensor = replace(other_tensor, old => new)
+                    tn.tensors[other_idx] = new_tensor
+                end
+            end
+
+            repeated_labels = replace(collect(labels(tensor)), old => new)
+            removed_label = filter(l -> l != old, labels(tensor))
+
+            data = EinCode((String.(repeated_labels),),[String.(removed_label)...])(tensor)
```
Comment (mofeing) [inline src/Transformations.jl:82]:
```suggestion
            # TODO rewrite with `Tensors` when it supports it
            # extract diagonal
            data = EinCode((String.(repeated_labels),),[String.(removed_label)...])(tensor)
```

**Pull request merged**